addon/components/tri-state.js

```diff
@@ -1,281 +1,193 @@
+    const willRenderLoadingComponent = this.get('promise.isPending') && !canShowLastSuccessful;
+    const willRenderErrorComponent = this.get('promise.isRejected') &&
+                                     !canShowLastSuccessful &&
+                                     !this.get('forceResolveAll');
+    const willRenderSuccessComponent = this.get('promise.isFulfilled') ||
+                                       (this.get('promise.isPending') && canShowLastSuccessful);
 
     return {
       error: {
-        isActive: this.get('isError'),
-        component: showError ? this.get('errorComponent') : noopComponent,
-        data: this.get('taskInstance.error'),
+        isActive: this.get('promise.isRejected'),
+        component: willRenderErrorComponent ? this.errorComponent : this.noopComponent,
+        data: willRenderErrorComponent ? this.get('promise.reason') : null,
       },
       success: {
-        isActive: this.get('isSuccess'),
-        component: showSuccess ? this.get('successComponent') : noopComponent,
-        data: this.get('lastSuccessValue'),
+        isActive: this.get('promise.isFulfilled'),
+        component: willRenderSuccessComponent ? this.successComponent : this.noopComponent,
+        data: willRenderSuccessComponent ? this.get('_lastResolvedData') : null,
       },
       loading: {
-        isActive: this.get('isLoading'),
-        component: showLoading ? this.get('loadingComponent') : noopComponent,
-        data: null,
+        isActive: this.get('promise.isPending'),
+        component: willRenderLoadingComponent ? this.loadingComponent : this.noopComponent,
       },
     };
   }),
 
   /**
-   * Set defaults and request data once we have attributes
+   * Helper function to resolve promises based on how they are provided
+   * @param {Promise|Array|Object} promises The promises to resolve
+   * @return {Promise} Promise which resolves when all promises have resolved
+   * @private
+   */
+  _resolvePromises(promises) {
+    if (promises instanceof Promise) {
+      return promises;
+    }
+
+    if (Array.isArray(promises)) {
+      if (this.get('forceResolveAll')) {
+        return RSVP.allSettled(promises);
+      }
+      return RSVP.all(promises);
+    }
+
+    if (typeof promises === 'object') {
+      if (this.get('forceResolveAll')) {
+        return RSVP.hashSettled(promises);
+      }
+      return RSVP.hash(promises);
+    }
+
+    return promises;
+  },
+
+  /**
+   * Set defaults and resolve `promise` once we have attributes
    */
   didReceiveAttrs() {
     this._super(...arguments);
 
     /**
-     * One or more actions that each return a promise.
-     * @type {Function|Object|Array}
+     * Cached data from the last fulfilled promise to use when `showLastSuccessful` is true
+     * @type {*}
+     * @private
      */
-    this.dataActions = this.getWithDefault('dataActions', noop);
+    this._lastResolvedData = this._lastResolvedData || false;
```

Comment:
Let's not mix types. If there is no data I would prefer null to be used rather than a boolean